### PR TITLE
feat: Table widget V2 Inline editing validation for plain text & number column

### LINKED ISSUE #13670: [Feature]: Validations for inline editing in input widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Add validations to the input widget within the table

### Why should this be worked on?

Completes the input inline editing flow

### PR DESCRIPTION
**Pull request by sbalaji1192**
## Description

Now editable column has the option to validate the input text before saving.

Fixes #13670

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Cypress test cases were added to verify the feature

Table widget inline editing validation functionality
  1. should check that validation only appears when editable enabled
  2. should check that validation only appears for plain text and number
  3. should check that regex, valid & required appear for the plain text column
  4. should check that min, max, regex, valid & required appear for the number column
  5. should check the validation property for the plain text column
     a. Regex
     b. Valid
     c. Required
  6. should check validation property for number column
     a. Min
     b. Max
  7. should check the error message property
  8. should check the editable cell actions when there is a validation error
     a. save should only work when there is no error
     b. discard should only work when there is no error
  9. should check that the save/discard button is disabled when there is a validation error

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by chandannkumar** [comment]
Observation:
1. Regex validation is not working for Number column type


2. Valid field is not working for Number type column
https://www.loom.com/share/aa6a2da4ae004ab1923b645b4caa2d9f

3. Error on the Table cell can seen outside the canvas as well
https://www.loom.com/share/cebd81a7029d46579472d4137fd285c7

4. There is no auto-suggestion for configuring valid field on Number or Plain text column type.
Like using **{{editedValue.length < 10}}**  no auto-suggestion was suggesting these

File: app/client/src/widgets/TableWidgetV2/widget/index.tsx [hunk 1]
```diff
@@ -1093,6 +1098,20 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
     return "TABLE_WIDGET_V2";
   }
 
+  getColumnByAlias(alias: string) {
+    const { primaryColumns } = this.props;
+
+    if (primaryColumns) {
+      const column = Object.values(primaryColumns).find(
+        (column) => column.alias === alias,
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/widget/index.tsx:1106]: `alias` can have non-ascii values right?
If so then can we do `column[alias]`?

Comment (ashit-rath) [reply app/client/src/widgets/TableWidgetV2/widget/index.tsx:1106]: Also is the method `getColumnByAlias` being used anywhere? Can't seem to find it

File: app/client/src/widgets/TableWidgetV2/widget/index.tsx [hunk 2]
```diff
@@ -1519,10 +1542,11 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
             columnType={column.columnType}
             compactMode={compactMode}
             disabledEditIcon={
-              this.props.inlineEditingSaveOption ===
+              (this.props.inlineEditingSaveOption ===
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/widget/index.tsx:1545]: Can we put this condition in a variable? it feels quite long to be part of component render.

File: app/client/src/widgets/TableWidgetV2/widget/index.tsx [hunk 3]
```diff
@@ -543,7 +546,8 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
      */
     if (isTableDataModified) {
       this.props.updateWidgetMetaProperty("transientTableData", {});
-      this.props.updateWidgetMetaProperty("editableCell", {});
+      this.props.updateWidgetMetaProperty("editableCell", defaultEditableCell);
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/widget/index.tsx:549]: Can we call `clearEditableCell` here?

File: app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/Validations/Common.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import { ValidationTypes } from "constants/WidgetValidation";
+
+export default [
+  {
+    helpText:
+      "Adds a validation to the cell value which displays an error on failure",
+    propertyName: "validation.regex",
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/Validations/Common.ts:7]: Can we have `propertyName` as the first key of the property objects? It helps finding the code faster.

**Review by ashit-rath — COMMENTED** [review]
Couple of things I noticed while trying out the DP

- The border here looks a bit weird to me as if the right and top border is thinner than the left and bottom ones.  Can you verify once?
Screenshot 2022-09-07 at 10 06 25 AM

- The validation sort of didn't work for me, is my binding incorrect here?
Screenshot 2022-09-07 at 10 12 48 AM

File: app/client/src/widgets/TableWidgetV2/widget/index.tsx (hunk 1 above)
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/widget/index.tsx:1106]: Stale code removed it.

File: app/client/src/widgets/TableWidgetV2/widget/derived.js [hunk 5]
```diff
@@ -587,4 +587,88 @@ export default {
...
+
+    if (editedColumn && editedColumn.validation) {
+      const validation = editedColumn.validation;
+
+      /* General validations */
+      if (!validation.isColumnEditableCellRequired && value === "") {
+        return true;
+      } else if (
+        (!_.isNil(validation.isColumnEditableCellValid) &&
+          !validation.isColumnEditableCellValid) ||
```
[the remark this answers is not in the document]
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/widget/derived.js:642]: We are making sure that isColumnEditableCellValid is not null or undefined before using that in negation.

File: app/client/src/widgets/TableWidgetV2/constants.ts [hunk 6]
```diff
@@ -78,14 +79,15 @@ export interface TableWidgetProps extends WidgetProps, WithMeta, TableStyles {
...
+  isEditableCellValid: boolean;
 }
 
 export const getCurrentRowBinding = (
   entityName: string,
   userInput: string,
   withBinding = true,
 ) => {
-  let rowBinding = `${entityName}.sanatizedTableData.map((currentRow) => ( ${userInput}))`;
+  let rowBinding = `${entityName}.processedTableData.map((currentRow, currentIndex) => ( ${userInput}))`;
```
[the remark this answers is not in the document]
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/constants.ts:90]: it's only used in the test. Let me move this to the test file.

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/InlineCellEditor.tsx [hunk 7]
```diff
@@ -122,9 +131,26 @@ export function InlineCellEditor({
   );
 
   const onTextChange = useCallback(
-    (data: string) => {
+    (inputValue: string) => {
```
[the remark this answers is not in the document]
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/InlineCellEditor.tsx:134]: Here the value will be coming from the input component which will always be a string.

File: app/client/src/components/editorComponents/ErrorTooltip.tsx [hunk 8]
```diff
@@ -38,9 +38,31 @@ interface Props {
   message: string;
   children: JSX.Element;
   customClass?: string;
+  preventOverflow?: string;
```
[the remark this answers is not in the document]
Comment (sbalaji1192) [reply app/client/src/components/editorComponents/ErrorTooltip.tsx:41]: This value is gettings passed to a component within a compoenet. So I guess we stick with the selector. If in the future we conditionally render the tooltip. we can avoid the unnecessary querySelector call if we let the errortooltip call it. Thoughs?

**Comment by sbalaji1192** [comment]
@ashit-rath Regarding your two points
1. I just zoomed in and had a closer look. the border seems to be fine. Can you try once as well?
2. Binding is incorrect you should use `editedValue` instead `currentRow.step`. while entering value currentRow.step value points to older value.

File: app/client/src/components/editorComponents/ErrorTooltip.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (ashit-rath) [reply app/client/src/components/editorComponents/ErrorTooltip.tsx:41]: Okay fair enough.

Can we modify the prop name to something that represent we need a selector here? Right now `preventOverflow` doesn't feel very intuitive.

**Comment by laveena-en** [comment]
A couple of observations here:

Can we have the error toast to read "This field is required" instead of invalid text when a field is set to required. That would be consistent with the other widgets and differentiate from an actual invalid input.



Invalid values are being allowed even when the valid condition is set https://www.loom.com/share/26d45749c5c4467f90be537d70634191

The error text is seen on the canvas on re-sizing https://www.loom.com/share/4b607ba78f6f424e9b73552cec075e4c

**Review by ashit-rath — APPROVED** [review]

**Comment by sbalaji1192** [comment]
@laveena-en 
#1 - fixed
#2 - not reproducible
#3 - not an issue. checked with @dilippitchika . we can ignore this.

**Comment by laveena-en** [comment]
LGTM!

**Review by ashit-rath — APPROVED** [review]

**Pull request merged**